server/services/caseListService.js

```diff
@@ -25,40 +33,74 @@ module.exports = function createCaseListService(nomisClientBuilder, roService, l
     }
   }
 
-  function getROCaseList(username, token) {
-    return async () => {
-      const deliusIds = await licenceClient.getDeliusUserName(username)
-
-      if (!Array.isArray(deliusIds) || deliusIds.length < 1 || isEmpty(deliusIds[0].staff_id)) {
-        logger.error(`No delius user ID for nomis ID '${username}'`)
-        return {
-          hdcEligible: [],
-          message: 'Delius username not found for current user',
-        }
-      }
+  /**
+   * Assume username is assigned to an RO.  Look up this user in the local db (staff_ids table) and take the staff code.
+   * Alternatively if a unique staff code cannot be found in this db, ask Delius for a staff code.
+   *
+   * The code below is an ( ugly bodge / elegant solution ) that breaks the type contract on Result.orElse():
+   * staffCodeFromDB is a Result<string,string>, but getStaffCodeFromDelius() returns a Promise<Result<string,string>>
+   * This means that the .orElse() will return either a Result<string,string> or a Promise<Result<string,string>>
+   * Fortunately Promise.then() flattens nested Promises so that it all 'just works', but there should be a better
+   * way to do this.
+   *
+   * Also, how do you add type annotations to curried functions???
+   */
+  const getROCaseList = (username, token) => () =>
+    R.pipeWith(R.then, [
+      getStaffCodeFromDb,
+      staffCodeFromDb => staffCodeFromDb.orElse(() => getStaffCodeFromDelius(username)),
+      sc => sc.map(getOffendersForStaffCode(token)).match(R.identity, message => ({ hdcEligible: [], message })),
+    ])(username)
+
+  /**
+   * @param {string} username
+   * @returns {Promise<Result<string, string>>}
+   */
+  const getStaffCodeFromDb = async username => {
+    const deliusIds = await licenceClient.getDeliusUserName(username)
+
+    return validateDeliusIds(deliusIds)
+      .flatMap(getDeliusId)
+      .map(id => id.staff_id.toUpperCase())
+  }
 
-      if (deliusIds.length > 1) {
-        logger.error(`Multiple delius user ID for nomis ID '${username}'`)
-        return {
-          hdcEligible: [],
-          message: 'Multiple Delius usernames found for current user',
-        }
-      }
+  /**
+   * @param {string} username
+   * @returns {Promise<Result<string, string>>}
+   */
+  const getStaffCodeFromDelius = async username => {
+    const staffDetailsResult = await getStaffDetailsFromDelius(username)
+
+    return staffDetailsResult.flatMap(staffDetails => {
+      const staffCode = R.prop('staffCode', staffDetails)
```

Comment:
would avoid using `R.prop` as that returns `any` and suggests that staffDetails could potentially be null when it can't be here?

Reply:
Good point!